pkg/controllers/internalmembercluster/v1beta1/member_controller.go

```diff
@@ -156,20 +272,183 @@ func (r *Reconciler) stopAgents(ctx context.Context, imc *clusterv1beta1.Interna
+	meta.SetStatusCondition(&imc.Status.Conditions, metav1.Condition{
+		Type:               string(clusterv1beta1.ConditionTypeClusterPropertyCollectionSucceeded),
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		ObservedGeneration: imc.GetGeneration(),
+	})
+}
+
+// reportPropertyProviderStartedCondition reports the condition of whether a property provider
+// has been started successfully.
+func reportPropertyProviderStartedCondition(imc *clusterv1beta1.InternalMemberCluster, status metav1.ConditionStatus, reason, message string) {
+	meta.SetStatusCondition(&imc.Status.Conditions, metav1.Condition{
+		Type:               string(clusterv1beta1.ConditionTypeClusterPropertyProviderStarted),
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		ObservedGeneration: imc.GetGeneration(),
+	})
+}
+
+// reportClusterPropertiesWithPropertyProvider collects the latest cluster properties from the property provider.
+func (r *Reconciler) reportClusterPropertiesWithPropertyProvider(ctx context.Context, imc *clusterv1beta1.InternalMemberCluster) error {
+	// Check if there are queued calls that have not returned yet.
+	ticket := r.propertyProviderCfg.queuedPropertyCollectionCalls.Add(1)
+	if ticket > int32(maxQueuedPropertyCollectionCalls) {
+		// There are too many on-going calls to the property provider; report the failure
+		// and skip this collection attempt.
+		reportPropertyProviderCollectionCondition(imc,
+			metav1.ConditionFalse,
+			ClusterPropertyCollectionFailedTooManyCallsReason,
+			ClusterPropertyCollectionFailedTooManyCallsMessage,
+		)
+		err := fmt.Errorf("too many on-going calls to the property provider; skipping this collection attempt")
+		klog.ErrorS(err, "Failed to collect cluster properties", "InternalMemberCluster", klog.KObj(imc))
+		return err
+	}
+
+	// Collect the latest cluster properties from the property provider.
+	//
+	// If the property provider fails to return the latest cluster properties within the deadline,
+	// report the failure as a condition.
+	childCtx, childCancelFunc := context.WithDeadline(ctx, time.Now().Add(propertyProviderCollectionDeadline))
+	// Cancel the context any way to avoid leaks.
+	defer childCancelFunc()
+
+	// Note that the property provider might not be able to complete the reporting within the
+	// deadline, and the Collect() call, despite our best efforts, might not terminate as the
+	// context exits; to avoid the corner case where too many routines are being created without the
+	// results being collected, Fleet will limit the number of on-going calls to the property
+	// provider.
+	collectedCh := make(chan struct{})
+
+	var res propertyprovider.PropertyCollectionResponse
+	go func() {
+		defer close(collectedCh)
+		defer r.propertyProviderCfg.queuedPropertyCollectionCalls.Add(-1)
```

Comment:
same question as the above, why use async when there is only one goroutine? the main thread is blocking anyway. What's the benefit?

Reply:
Yeah, similarly, Collect() is implemented externally, and on the caller side there's no guarantee that the call would not block and get cancelled promptly if it timed out.